drivers/rn2xx3/rn2xx3_internal.c

```diff
@@ -0,0 +1,317 @@
+        *pos++ = hex_chars[(*byte_array >> 4) & 0xF];
+        *pos++ = hex_chars[(*byte_array++) & 0xF];
+    }
+    *pos = '\0';
+}
+
+void rn2xx3_set_internal_state(rn2xx3_t *dev, uint8_t state)
+{
+    if (dev->int_state == RN2XX3_INT_STATE_SLEEP || dev->int_state == state) {
+        return;
+    }
+
+    DEBUG("[rn2xx3] new state: %d\n", state);
+    dev->int_state = state;
+}
+
+int rn2xx3_write_cmd(rn2xx3_t *dev)
+{
+    DEBUG("[rn2xx3] CMD: %s\n", dev->cmd_buf);
+
+    if (dev->int_state == RN2XX3_INT_STATE_SLEEP) {
+        DEBUG("[rn2xx3] ABORT: device is in sleep mode\n");
+        return RN2XX3_ERR_SLEEP_MODE;
+    }
+
+    uint8_t prev_state = dev->int_state;
+    rn2xx3_set_internal_state(dev, RN2XX3_INT_STATE_CMD);
+
+    mutex_lock(&(dev->cmd_lock));
+    uart_write(dev->p.uart, (uint8_t*)dev->cmd_buf, strlen(dev->cmd_buf));
+    _terminate_command(dev);
+
+    if (rn2xx3_wait_response(dev) == RN2XX3_TIMEOUT) {
+        mutex_unlock(&(dev->cmd_lock));
+        return RN2XX3_TIMEOUT;
+    }
+
+    mutex_unlock(&(dev->cmd_lock));
+
+    rn2xx3_set_internal_state(dev, prev_state);
+
+    return rn2xx3_process_response(dev);
+}
+
+int rn2xx3_write_cmd_no_wait(rn2xx3_t *dev)
+{
+    DEBUG("[rn2xx3] CMD: %s\n", dev->cmd_buf);
+
+    if (dev->int_state == RN2XX3_INT_STATE_SLEEP) {
+        DEBUG("[rn2xx3] ABORT: device is in sleep mode\n");
+        return RN2XX3_ERR_SLEEP_MODE;
+    }
+
+    mutex_lock(&(dev->cmd_lock));
+    uart_write(dev->p.uart, (uint8_t*)dev->cmd_buf, strlen(dev->cmd_buf));
+    _terminate_command(dev);
+
+    DEBUG("[rn2xx3] RET: %s\n", dev->resp_buf);
+
+    mutex_unlock(&(dev->cmd_lock));
+
+    return rn2xx3_process_response(dev);
+}
+
+int rn2xx3_wait_response(rn2xx3_t *dev)
+{
+    if (_wait_reply(dev, RESP_TIMEOUT_SEC)) {
+        return RN2XX3_TIMEOUT;
+    }
+
+    DEBUG("[rn2xx3] RESP: %s\n", dev->resp_buf);
+
+    return RN2XX3_OK;
+}
+
+int rn2xx3_wait_reply(rn2xx3_t *dev, uint8_t timeout)
+{
+    if (_wait_reply(dev, timeout)) {
+        return RN2XX3_REPLY_TIMEOUT;
+    }
+
+    DEBUG("[rn2xx3] REPLY: %s\n", dev->resp_buf);
+
+    return rn2xx3_process_reply(dev);
+}
+
+void rn2xx3_cmd_start(rn2xx3_t *dev)
+{
+    rn2xx3_set_internal_state(dev, RN2XX3_INT_STATE_CMD);
+    DEBUG("[rn2xx3] CMD: %s", dev->cmd_buf);
+    mutex_lock(&(dev->cmd_lock));
+    uart_write(dev->p.uart, (uint8_t*)dev->cmd_buf, strlen(dev->cmd_buf));
+}
+
+void rn2xx3_cmd_append(rn2xx3_t *dev, uint8_t *payload, uint8_t payload_len)
+{
+    char payload_str[2];
+    rn2xx3_bytes_to_hex(payload, payload_str, payload_len);
+    DEBUG("%s", payload_str);
+    uart_write(dev->p.uart, (uint8_t*)payload_str, payload_len * 2);
+}
+
+int rn2xx3_cmd_finalize(rn2xx3_t *dev)
+{
+    DEBUG("\n");
+
+    uart_write(dev->p.uart, (uint8_t*)"\r\n", 2);
+    if (rn2xx3_wait_response(dev) == RN2XX3_TIMEOUT) {
+        rn2xx3_set_internal_state(dev, RN2XX3_INT_STATE_IDLE);
+        mutex_unlock(&(dev->cmd_lock));
+        return RN2XX3_TIMEOUT;
```

Comment:
By introducing a retval helper variable you can make this function to have a single exit. The single exit point can do the mutex_unlock. And the code readability will not suffer, I think.